datafusion/physical-optimizer/src/push_down_filter.rs

```diff
@@ -363,46 +362,139 @@ use datafusion_physical_plan::ExecutionPlan;
 /// [`AggregateExec`]: datafusion_physical_plan::aggregates::AggregateExec
 #[derive(Debug)]
 pub struct PushdownFilter {}
-
 impl Default for PushdownFilter {
     fn default() -> Self {
         Self::new()
     }
 }
 
-pub type FilterDescriptionContext = PlanContext<FilterDescription>;
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum FilterPushdownState {
+    NoChildren,
+    Unsupported,
+    Supported,
+}
+
+fn push_down_filters(
+    node: &Arc<dyn ExecutionPlan>,
+    parent_filters: Vec<Arc<dyn PhysicalExpr>>,
+    config: &ConfigOptions,
+) -> Result<FilterPushdownPropagation<Arc<dyn ExecutionPlan>>> {
+    let pushdown_plan = node.gather_filters_for_pushdown(&parent_filters, config)?;
+    let children = node
+        .children()
+        .into_iter()
+        .map(Arc::clone)
+        .collect::<Vec<_>>();
+    let mut parent_pushdown_result =
```

Comment:
one idea to make this code easier to follow might be to break it up into multiple functions. Maybe even a struct or something that held the key state

```rust
pub struct SingleNodePushdown {
  pushdown_plan: ....
  new_children: Vec<>
}
...
```